### DISCUSSION TITLE
ESP32: I (871) + Flashing AI-deck

### DISCUSSION THREAD

**Discussion by NSombekke** [bitcraze/discussions#173] [Q&A]
Dear community,

I have been working with the Crazyflie 2.0 for the last couple of weeks and I finally got my hands on the AI-deck 1.1 (Revision D). I followed the getting started with AI-deck tutorial and updated the firmware of both the Crazyflie and AI-deck (through the cfclient bootloader on the bitcrazeVM) to version 2022.05. This actually took around 25 minutes (not the indicated 3 minutes).

When reconnecting the Crazyflie to the cfclient the console showed (besides other things):
`ESP32: I (871) SYS: Initialized`
`ESP32: I (881) DISCOVERY: Hostname is aideck-E0E3DC`
The first one differs from `ESP32: I (910) SYS: Initialized` which can be found in the getting started tutorial, why this difference and is this a problem?

I continued with the tutorial as it did seem to work (apart from the difference mentioned above). Because I have revision D of the AI-deck I assumed I did not have to flash the bootloader on the GAP8 seperately but when continuing with the Flash Wifi example I faced new problems, similar to #57. The proposed solution in #57 did not seem to work as flashing kept being stuck at 0 or 2 percent. After being stuck for a while the Crazyflie seems to restart (the usual beeping sound and spinning rotors). Does anyone have a solution for this as it has been frustrating me for days now.

**Comment by mjschock** [comment]
I don't have an answer for this but figured I add my experience as it's similar. I also see a discrepancy with that 910 number from the tutorial.  I've got the Crazyflie 2.1 and the AI-deck 1.1.  I'm not sure how to determine the revision, though.  My output looks like this:

```
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:8c606d1f8f89 (2022.05) CLEAN
SYS: I am 0x20303743575350170029001E and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcAI for deck 0
AIDECK-ROUTER: Not setting up WiFi
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2022.0
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 16616 bytes
ESP32: I (494) SYS: Initialized
ESP32: I (504) DISCOVERY: Hostname is aideck-CC3858
ESP32: W (494) ROUTER: Cannot route from GAP8 [0x00] to [0x00]
STAB: Ready to fly.
```
Initially, when trying to flash the WiFi Example, it would fail arbitrarily at different percentages when flashing. I had the Flow v2 deck attached and ignored step 2, which says, "Make sure that only the AI-deck is attached to the Crazyflie, with no other deck." As soon as I removed the flow deck and retried the flash, it flashed to 99%.  And it does this repeatedly if I retry, always stopping at 99%.

If I try `radio://0/80/2M` instead of `radio://0/80/2M/E7E7E7E7E7` as suggested in #57, it also stops at 99%.

I ended up stepping through the code during the flash and found that the flash actually reaches 100% and then exists with `False` here:  https://github.com/bitcraze/crazyflie-lib-python/blob/8b6f740eb131083d024f9129073b9b8bf0f735f2/cflib/crazyflie/mem/__init__.py#L225.  I.e., `new_progress` is at 99% *before* the call to `self._write_new_chunk()` and recalculating `new_progress` after that call shows 100%.  I tried returning `True` in that specific case, and it finished the flash.  Here's the change I made:  https://github.com/mjschock/crazyflie-lib-python/commit/4fd0649ca605cf082bcb04df8ca3a002de31aa63.

Unfortunately, the `WiFi streaming example` SSID doesn't surface, and connecting to it as if it's hidden doesn't work either.

**Comment by knmcguire** [comment]
Hi @NSombekke,

Sorry that you are having these problems. I'm just asking some extra questions to get some clarification

- With the wifi example flashing, you are talking about [this step](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#flash-wifi-example) right?, not [the continued guides](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/test-functions/wifi-streamer/) one? 
- Could you send the debug information when connected to the crazyflie to the cflclient at 'help'>'about'>'debug' tab.

**Reply by NSombekke** [reply]
Hi Kimberly, thanks for your response!

Yes I am talking about the [first one](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#flash-wifi-example). 

This is the debug information I got from the cflcient:

```
Cfclient
Cfclient version: 2022.5.post71+c10cd24
System: linux
Python: 3.8.10
Qt: 5.15.2
PyQt: 5.15.7

Interface status
radio: Crazyradio version 0.53
UsbCdc: No information available
udp: None
prrt: No information available

Input readers
linuxjsdev (2 devices connected)

Input devices
None

Crazyflie
Connected: radio://0/80/250K
Firmware: 8c606d1f8f89 (CLEAN)

Decks found
bcAI: rev=C1, adr=0D1A65DC0000007C

Sensors found
BMP388: 0
AK8963: 1
LPS25H: 1

Sensors tests
MPU6500: 1
AK8963: 1
LPS25H: 1
```

I notice that it indicates the AI-deck has rev=C1, which is strange as the AI-deck has both rev D marked on the deck itself and on the bag it was delivered in. Could this potentially be the problem?

**Reply by knmcguire** [reply]
Thanks! 

the one about the revc1 while revD is printed on the board... I think this went wrong in production that the right REV number was not properly stored in the one wire memory of the deck. But the firmware should be up to date so you should have the bootloader.

I might have an idea what is going wrong in your setup so I'll propose a potential solution in a new thread

**Comment by knmcguire** [comment]
So, first of all, the 25 min over air flash time of the esp firmware is quite worrying but I see why that is from your debug information:

`Connected: radio://0/80/250K
`
250Kbit/s is a very low bitrate and I'm not sure that the cfloader or the cfclient is tested on that with the aideck. Usually the newer crazyflies come preconfigured with 2Mbit/s but since you are using a crazyflie 2.0 that is probably not the case.

Could you change the radio properties [with these instructions](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/#firmware-configuration)? Select 2MBit/s and perhaps also change the radio address to be E7E7E7E701 to be sure that we don't have the issue described int #57. Then restart the crazyflie and go through [the entire tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/) again.

Also just to be sure, let's try the latest release of the cfclient, as your debug information indicate that you are running from the latest source, `Cfclient version: 2022.5.post71+c10cd24`. This should be `Cfclient version: 2022.5` as the latest version might be unstable.  I don't think this is the issue but you never know. 

```
pip3 uninstall cflib cfclient
pip3 install cflient
```

Let us know how this goes

**Reply by NSombekke** [reply]
Even though the air flash time slightly sped up to around 15 minutes, the wifi-streamer flash example is sadly still stuck at 0 or 2 percent.
```
cfloader flash aideck_gap8_wifi_img_streamer_with_ap.bin deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E701
Reset to bootloader mode ...
Could not save cache, no writable directory
Could not save cache, no writable directory
Skipping bcAI:esp, not in the target list
Deck bcAI:gap8, reset to bootloader
| 0% Writing to bcAI:gap8 deck memory
/ 1% Writing to bcAI:gap8 deck memory
- 2% Writing to bcAI:gap8 deck memory
```
I have tried it twice now with the radio://0/80/2M/E7E7E7E701 adress (could not find the Crazyflie when leaving the E7E7E7E701 part out).
The first time it stopped for a few seconds on 0%, the Crazyflie restarted (beeping sound, spinning proppelers) and it continued to 2% where it stopped indefinitely.
The second time it stopped at 0%, the Crazyflie restarted and showed a self-test fail (five short red pulses, front right led). 
Is the restart of the Crazyflie normal behavior?

**Reply by knmcguire** [reply]
Arg I was hoping the 2Kbit was the limited factor, but it seems it is not...

The self-test fail is worrisome, especially when it's from the latest release. But I have another idea that you can try out which I'll propose in a second in a new thread

**Comment by knmcguire** [comment]
Not sure what is going one, but one of the things you can try some firmwares first, so hopefully we can avoid you using a programmer to fix things:

- Install the latest STM firmware version from source via these instructions: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/
- Install the latest ESP firmware from source via these instructions: https://github.com/bitcraze/aideck-esp-firmware:
  - First use docker to build the ESP firmware (this will take a while): https://github.com/bitcraze/aideck-esp-firmware#flash-using-the-builder-image
  - Don't use the docker instructions to flash it (that requires a programmer) but use 'flash using the cfloader: https://github.com/bitcraze/aideck-esp-firmware#flash-using-the-cfloader

Then try to flash the binary for the wifi streamer again.

In the mean time a question: Do you have an olimex or jlink programmer  at your disposal just in case this doesn't work?

**Reply by NSombekke** [reply]
Thanks, I will try and let you know. 
I don't have a programmer here at the moment but I can ask my Uni if they have got one lying around that I can borrow.

**Reply by NSombekke** [reply]
Installation of the STM firmware succeeded without any problems, but when installing the ESP firmware I face a problem with building the firmware in docker with this command:
```
`docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/builder /bin/bash -c "make"`
```
When it is almost finished it stops the process and shows the following:
```
docker: failed to register layer: Error processing tar file(exit status 1): write /new_home/.espressif/tools/riscv32-esp-elf/esp-2021r1-8.4.0/riscv32-esp-elf/riscv32-esp-elf/lib/rv32im/ilp32/libstdc++.a: no space left on device.
See 'docker run --help'.
```
I am currently working in the virtual machine provided by Bitcraze, I already tried a clean install but with no success.

**Reply by knmcguire** [reply]
no space left on device means that there is no more diskspace left on your virtual machine. You should add more diskspace, which hopefully will be possible as a setting from your host program.

**Comment by knmcguire** [comment]
In a previous thread you mentioned that you use a virtual machine, which is displaying communication performances if not optimized for use. There are some tips to improve that: https://github.com/bitcraze/bitcraze-vm

The best would be to do aideck development from a computer with a native ubuntu install. WSL2 (windows subsystem for linux) on window10/11 works too actually if you want some instructions for that.
